File: ios/RNToolsManager.m
//
//  RNToolsManager.m
//  bbn_yahoo_bid
//
//  Created by Bibian App RD on 2023/1/13.
//

// https://www.jianshu.com/p/6fc3f77e0213
#import "RNToolsManager.h"

@implementation RNToolsManager

@synthesize bridge = _bridge;

RCT_EXPORT_MODULE(RNToolsManager);
//  宣告給ReactNative使用，取得app版本號
RCT_EXPORT_METHOD(getAppVersion:(RCTPromiseResolveBlock)resolve  rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    NSString *version = [[[NSBundle mainBundle] infoDictionary] objectForKey:@"CFBundleShortVersionString"];// 取得目前版本號
    resolve(version);
  }
  @catch (NSException *exception) {
    reject(@"ios error getAppVersion", exception.description, nil);
  }
}
//  宣告給ReactNative使用，TappaySDK的initInstance
RCT_EXPORT_METHOD(TappayInitInstance:(nonnull NSNumber *)APP_ID APP_KEY:(NSString *)APP_KEY prod:(BOOL)prod resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    self.TappayManager = [TappayManager alloc];
    [self.TappayManager initInstance:APP_ID APP_KEY:APP_KEY prod:prod];
    
    resolve(@{
      @"systemOS": @"ios",
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"success":@ YES
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayInitInstance", exception.description, nil);
  }
}
//  宣告給ReactNative使用，TappaySDK的TPDCard
RCT_EXPORT_METHOD(TappaySetTPDCard:(NSString *)cardNumber dueMonth:(NSString *)dueMonth dueYear:(NSString *)dueYear CCV:(NSString *)CCV resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    TPDCardValidationResult *result = [self.TappayManager TPDCard:cardNumber dueMonth:dueMonth dueYear:dueYear CCV:CCV];
    NSString *cardType = [self.TappayManager cardTypeToString:result.cardType];

    resolve(@{
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"systemOS": @"ios",
      @"cardNumber":cardNumber,
      @"dueMonth":dueMonth,
      @"dueYear":dueYear,
      @"CCV":CCV,
      @"isCardNumberValid":[NSNumber numberWithBool:result.isCardNumberValid],
      @"isExpiryDateValid":[NSNumber numberWithBool:result.isExpiryDateValid],
      @"isCCVValid":[NSNumber numberWithBool:result.isCCVValid],
      @"cardType":cardType,
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappaySetTPDCard", exception.description, nil);
  }
}
//  宣告給ReactNative使用，TappaySDK的createTokenWithGeoLocation
RCT_EXPORT_METHOD(TappayGetDirectPayPrime:(NSString *)geoLocation resolver:(RCTPromiseResolveBlock)resolve  rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    [self.TappayManager getDirectPayPrime:geoLocation resolver:resolve rejecter:reject];
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayGetDirectPayPrime", exception.description, nil);
  }
}
//  宣告給ReactNative使用，TappaySDK的getDeviceId
RCT_EXPORT_METHOD(TappayGetDeviceId:(RCTPromiseResolveBlock)resolve  rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
   resolve([self.TappayManager getDeviceId]);
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayGetDeviceId", exception.description, nil);
  }
}

//  宣告給ReactNative使用，對應android版的TappayGooglePayInit方法
RCT_EXPORT_METHOD(TappayGooglePayInit:(NSString *)merchantName resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  // reject(@"ios error TappayGooglePayInit", @"ios not support GooglePay", nil);
  resolve(@{
    @"systemOS": @"ios",
    @"tappaySDKVersion": self.TappayManager.SDKVersion,
    @"isReadyToPay": @NO,
    @"msg": @"ios not support GooglePay"
  });
}

//  宣告給ReactNative使用，對應android版的TappayHandlerGooglePay方法
RCT_EXPORT_METHOD(TappayGetGooglePayPrime:(NSString *)merchantName resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  reject(@"ios error TappayGetGooglePayPrime", @"ios not support GooglePay", nil);
}

// 宣告給ReactNative使用的function名稱，對TappaySDK的TPDapplePay做初始化
RCT_EXPORT_METHOD(TappayAapplePayInit:(NSString *)merchantName merchantId:(NSString *)merchantId countryCode:(NSString *)countryCode currencyCode:(NSString *)currencyCode resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    [self.TappayManager applePayInit:merchantName merchantId:merchantId countryCode:countryCode currencyCode:currencyCode resolver:resolve rejecter:reject];
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayAapplePayInit", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK的TPDapplePay的canMakePayments
RCT_EXPORT_METHOD(TappayIsApplePayAvailable:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    BOOL result = [self.TappayManager isApplePayAvailable];
    resolve(@{
      @"systemOS": @"ios",
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"isReadyToPay": @(result)
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayIsApplePayAvailable", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDapplePay的付款方法
RCT_EXPORT_METHOD(TappayGetApplePayPrime:(NSString *)amount resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    [self.TappayManager getApplePayPrime:amount resolver:resolve rejecter:reject];
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayGetApplePayPrime", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDlinePay的handleURL方法
RCT_EXPORT_METHOD(TappayLinePayHandleURL:(NSString *)openUri resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    bool success = [self.TappayManager linePayHandleURL:openUri];
    resolve(@{
      @"systemOS": @"ios",
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"openUri": openUri,
      @"success": @(success)
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayIsLinePayAvailable", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDlinePay的isLinePayAvailable方法
RCT_EXPORT_METHOD(TappayIsLinePayAvailable:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    bool isReadyToPay = [self.TappayManager isLinePayAvailable];
    resolve(@{
      @"systemOS": @"ios",
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"isReadyToPay": @(isReadyToPay)
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayIsLinePayAvailable", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDlinePay的初始化方法
RCT_EXPORT_METHOD(TappayLinePayInit:(NSString *)linePayCallbackUri resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    bool isLinePayAvailable = [self.TappayManager linePayInit:linePayCallbackUri];
    resolve(@{
      @"systemOS": @"ios",
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"isReadyToPay": @(isLinePayAvailable),
      @"linePayCallbackUri": linePayCallbackUri
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayLinePayInit", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDlinePay的getPrime
RCT_EXPORT_METHOD(TappayGetLinePayPrime:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    [self.TappayManager getLinePayPrime:resolve rejecter:reject];
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayGetLinePayPrime", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDlinePay的redirect
RCT_EXPORT_METHOD(TappayLinePayRedirectWithUrl:(NSString *)paymentUrl resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    [self.TappayManager linePayRedirectWithUrl:paymentUrl resolver:resolve rejecter:reject];
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayLinePayRedirectWithUrl", exception.description, nil);
  }
}

//  宣告給ReactNative使用，對應android版的TappaySamsungPayInit方法
RCT_EXPORT_METHOD(TappaySamsungPayInit:(NSString *)merchantName merchantId:(NSString *)merchantId currencyCode:(NSString *)currencyCode serviceId:(NSString *)serviceId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    resolve(@{
      @"systemOS": @"ios",
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"isReadyToPay": @NO,
      @"msg": @"ios not support SamsungPay"
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappaySamsungPayInit", exception.description, nil);
  }
}

//  宣告給ReactNative使用，對應android版的TappayGetSamsungPayPrime方法
RCT_EXPORT_METHOD(TappayGetSamsungPayPrime:(NSString *)itemTotalAmount shippingPrice:(NSString *)shippingPrice tax:(NSString *)tax totalAmount:(NSString *)totalAmount resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  reject(@"ios error TappayGetSamsungPayPrime", @"ios not support SamsungPay", nil);
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDJkoPay的isJkoPayAvailable
RCT_EXPORT_METHOD(TappayIsJkoPayAvailable:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    resolve(@{
      @"systemOS": @"ios",
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"isReadyToPay": @([self.TappayManager isJkoPayAvailable])
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayIsJkoPayAvailable", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDJkoPay的初始化方法
RCT_EXPORT_METHOD(TappayJkoPayInit:(NSString *)_jkoPayUniversalLinks resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    resolve(@{
      @"systemOS": @"ios",
      @"tappaySDKVersion": self.TappayManager.SDKVersion,
      @"isReadyToPay": @([self.TappayManager jkoPayInit:_jkoPayUniversalLinks])
    });
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayJkoPayInit", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDJkoPay取得prime方法
RCT_EXPORT_METHOD(TappayGetJkoPayPrime:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    [self.TappayManager getJkoPayPrime:resolve rejecter:reject];
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayGetJkoPayPrime", exception.description, nil);
  }
}

// 宣告給ReactNative使用的function名稱，對TappaySDK TPDJkoPay的跳轉付款方法
RCT_EXPORT_METHOD(TappayJkoPayRedirectWithUrl:(NSString *)paymentUrl resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
  @try {
    [self.TappayManager jkoPayRedirectWithUrl:paymentUrl resolver:resolve rejecter:reject];
  }
  @catch (NSException *exception) {
    reject(@"ios error TappayJkoPayRedirectWithUrl", exception.description, nil);
  }
}

@end

File: android/app/src/main/java/com/reactnativetappayhook/RNToolsManager.java
package com.reactnativetappayhook;

// import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
// import com.facebook.react.uimanager.IllegalViewOperationException;
// import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.WritableNativeMap;

import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import tech.cherri.tpdirect.api.TPDCardValidationResult;

// https://codeantenna.com/a/CRDAQIUluT
public class RNToolsManager extends ReactContextBaseJavaModule {
  TappayManager TappayManager;

  public RNToolsManager(ReactApplicationContext reactContext) {
    super(reactContext);
    TappayManager = new TappayManager(reactContext);
  }

  // 複寫getName方法定義Module名稱,在ReactNative呼叫時使用
  @Override
  public String getName() {
    return "RNToolsManager";
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的initInstance
  @ReactMethod
  public void TappayInitInstance(int APP_ID, String APP_KEY, Boolean prod, Promise promise) {
    try {
      TappayManager.initInstance(APP_ID, APP_KEY, prod);
      WritableNativeMap resultData = new WritableNativeMap();

      resultData.putString("systemOS", "android");
      resultData.putString("tappaySDKVersion", TappayManager.SDKVersion);
      resultData.putBoolean("success", true);

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappayInitInstance", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDCard
  @ReactMethod
  public void TappaySetTPDCard(String cardNumber, String dueMonth, String dueYear, String CCV,
      Promise promise) {
    try {
      TPDCardValidationResult result = TappayManager.TPDCard(cardNumber, dueMonth, dueYear, CCV);
      WritableNativeMap resultData = new WritableNativeMap();

      resultData.putString("systemOS", "android");
      resultData.putString("tappaySDKVersion", TappayManager.SDKVersion);
      resultData.putString("cardNumber", cardNumber);
      resultData.putString("dueMonth", dueMonth);
      resultData.putString("dueYear", dueYear);
      resultData.putString("CCV", CCV);
      resultData.putBoolean("isCardNumberValid", result.isCardNumberValid());
      resultData.putBoolean("isExpiryDateValid", result.isExpiryDateValid());
      resultData.putBoolean("isCCVValid", result.isCCVValid());
      resultData.putString("cardType", String.valueOf(result.getCardType()));

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappaySetTPDCard", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的createToken
  @ReactMethod
  public void TappayGetDirectPayPrime(String geoLocation, Promise promise) {
    try {
      TappayManager.getDirectPayPrime(geoLocation, promise);
    } catch (Exception e) {
      promise.reject("android error TappayGetDirectPayPrime", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的getRbaDeviceId
  @ReactMethod
  public void TappayGetDeviceId(Promise promise) {
    try {
      promise.resolve(TappayManager.getDeviceId());
    } catch (Exception e) {
      promise.reject("android error TappayGetDeviceId", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，對TappaySDK的TPDGooglePay做初始化
  @ReactMethod
  public void TappayGooglePayInit(String merchantName, Promise promise) {
    try {
      TappayManager.googlePayInit(merchantName, promise);
    } catch (Exception e) {
      promise.reject("android error TappayGooglePayInit", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDGooglePay的requestPayment
  @ReactMethod
  public void TappayGetGooglePayPrime(String totalPrice, String currencyCode, Promise promise) {
    try {
      TappayManager.getGooglePayPrime(totalPrice, currencyCode, promise);
    } catch (Exception e) {
      promise.reject("android error TappayGetGooglePayPrime", e);
    }
  }

  // 宣告給ReactNative使用，對應ios版的TappayAapplePayInit方法
  @ReactMethod
  public void TappayAapplePayInit(String merchantName, String merchantId, String countryCode, String currencyCode,
      Promise promise) {
    try {
      WritableNativeMap resultData = new WritableNativeMap();
      resultData.putBoolean("isReadyToPay", false);

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappayAapplePayInit", e);
    }
  }

  // 宣告給ReactNative使用，對應ios版的TappayIsApplePayAvailable方法
  @ReactMethod
  public void TappayIsApplePayAvailable(Promise promise) {
    try {
      WritableNativeMap resultData = new WritableNativeMap();
      resultData.putBoolean("isReadyToPay", false);

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappayIsApplePayAvailable", e);
    }
  }

  // 宣告給ReactNative使用，對應ios版的TappayHandlerApplePay方法
  @ReactMethod
  public void TappayGetApplePayPrime(String amount, Promise promise) {
    promise.reject("android error TappayGetApplePayPrime", "android not support ApplePay");
  }

  // 宣告給ReactNative使用的function名稱，對應ios版的TappayLinePayHandleURL方法
  @ReactMethod
  public void TappayLinePayHandleURL(String openUri, Promise promise) {
    try {
      WritableNativeMap resultData = new WritableNativeMap();
      resultData.putString("systemOS", "android");
      resultData.putString("tappaySDKVersion", TappayManager.SDKVersion);
      resultData.putString("openUri", openUri);
      resultData.putBoolean("success", false);

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappayIsLinePayAvailable", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDLinePay的isLinePayAvailable
  @ReactMethod
  public void TappayIsLinePayAvailable(Promise promise) {
    try {
      boolean result = TappayManager.isLinePayAvailable();
      WritableNativeMap resultData = new WritableNativeMap();
      resultData.putString("systemOS", "android");
      resultData.putString("tappaySDKVersion", TappayManager.SDKVersion);
      resultData.putBoolean("isReadyToPay", result);

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappayIsLinePayAvailable", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDLinePay的linePayInit
  @ReactMethod
  public void TappayLinePayInit(String linePayCallbackUri, Promise promise) {
    try {
      boolean result = TappayManager.linePayInit(linePayCallbackUri);
      WritableNativeMap resultData = new WritableNativeMap();
      resultData.putString("systemOS", "android");
      resultData.putString("tappaySDKVersion", TappayManager.SDKVersion);
      resultData.putBoolean("isReadyToPay", result);
      resultData.putString("linePayCallbackUri", linePayCallbackUri);

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappayLinePayInit", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDLinePay的getPrime
  @ReactMethod
  public void TappayGetLinePayPrime(Promise promise) {
    try {
      TappayManager.getLinePayPrime(promise);
    } catch (Exception e) {
      promise.reject("android error TappayGetLinePayPrime", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDLinePay的redirectWithUrl，並自動完成跳轉資料的監聽
  @ReactMethod
  public void TappayLinePayRedirectWithUrl(String paymentUrl, Promise promise) {
    try {
      TappayManager.linePayRedirectWithUrl(paymentUrl, promise);
    } catch (Exception e) {
      promise.reject("android error TappayLinePayRedirectWithUrl", e);
    }
  }

  @ReactMethod
  public void TappaySamsungPayInit(String merchantName, String merchantId, String currencyCode,
      String serviceId, Promise promise) {
    try {
      TappayManager.samsungPayInit(merchantName, merchantId, currencyCode, serviceId, promise);
    } catch (Exception e) {
      promise.reject("android error TappaySamsungPayInit", e);
    }
  }

  @ReactMethod
  public void TappayGetSamsungPayPrime(String itemTotalAmount, String shippingPrice, String tax, String totalAmount,
      Promise promise) {
    try {
      TappayManager.getSamsungPayPrime(itemTotalAmount, shippingPrice, tax, totalAmount, promise);
    } catch (Exception e) {
      promise.reject("android error TappayGetSamsungPayPrime", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDJkoPay的isJkoPayAvailable
  @ReactMethod
  public void TappayIsJkoPayAvailable(Promise promise) {
    try {
      boolean result = TappayManager.isJkoPayAvailable();
      WritableNativeMap resultData = new WritableNativeMap();
      resultData.putString("systemOS", "android");
      resultData.putString("tappaySDKVersion", TappayManager.SDKVersion);
      resultData.putBoolean("isReadyToPay", result);

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappayIsJkoPayAvailable", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDJkoPay的初始化方法
  @ReactMethod
  public void TappayJkoPayInit(String jkoPayUniversalLinks, Promise promise) {
    try {
      boolean result = TappayManager.jkoPayInit(jkoPayUniversalLinks);
      WritableNativeMap resultData = new WritableNativeMap();
      resultData.putString("systemOS", "android");
      resultData.putString("tappaySDKVersion", TappayManager.SDKVersion);
      resultData.putBoolean("isReadyToPay", result);

      promise.resolve(resultData);
    } catch (Exception e) {
      promise.reject("android error TappayJkoPayInit", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDJkoPay取得prime方法
  @ReactMethod
  public void TappayGetJkoPayPrime(Promise promise) {
    try {
      TappayManager.getJkoPayPrime(promise);
    } catch (Exception e) {
      promise.reject("android error TappayGetJkoPayPrime", e);
    }
  }

  // 宣告給ReactNative使用的function名稱，TappaySDK的TPDJkoPay的跳轉付款方法
  @ReactMethod
  public void TappayJkoPayRedirectWithUrl(String paymentUrl, Promise promise) {
    try {
      TappayManager.jkoPayRedirectWithUrl(paymentUrl, promise);
    } catch (Exception e) {
      promise.reject("android error TappayJkoPayRedirectWithUrl", e);
    }
  }
}